internal/tool/multi/deployer.go

```diff
@@ -356,6 +379,59 @@ func (h *handler) subscribeTo(req *protos.ActivateComponentRequest) error {
 	return h.envelope.UpdateRoutingInfo(target.routing(req.Component))
 }
 
+// VerifyClientCertificate implements the envelope.EnvelopeHandler interface.
+func (h *handler) VerifyClientCertificate(_ context.Context, req *protos.VerifyClientCertificateRequest) (*protos.VerifyClientCertificateReply, error) {
+	_, err := h.verifyCertificate(req.CertChain)
+	if err != nil {
+		return nil, err
+	}
+
+	// For now, return all components.
+	// TODO(spetrovic): Use the call graph to return the set of components
+	// that the client group is allowed to invoke methods on.
+	return &protos.VerifyClientCertificateReply{
+		Components: h.envelope.WeaveletInfo().Components,
+	}, nil
+}
+
+// VerifyServerCertificate implements the envelope.EnvelopeHandler interface.
+func (h *handler) VerifyServerCertificate(_ context.Context, req *protos.VerifyServerCertificateRequest) (*protos.VerifyServerCertificateReply, error) {
+	actual, err := h.verifyCertificate(req.CertChain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the expected group name for the target component.
+	expected, ok := h.colocation[req.TargetComponent]
+	if !ok {
+		expected = req.TargetComponent
+	}
+	if expected != actual {
+		return nil, fmt.Errorf("invalid server identity for target component %s: want %q, got %q", req.TargetComponent, expected, actual)
+	}
+	return &protos.VerifyServerCertificateReply{}, nil
+}
+
+func (h *handler) verifyCertificate(certChain [][]byte) (string, error) {
+	if n := len(certChain); n != 1 {
+		return "", fmt.Errorf("invalid cert chain length: want 1, got %d", n)
+	}
+	names, err := certs.VerifySignedCert(certChain[0], h.caCert)
+	if err != nil {
+		return "", fmt.Errorf("cannot verify the client cert: %w", err)
```

Comment:
Maybe reword to avoid using the word client since this function is used to verify client and server certs?